src/mlpack/core/data/save_image_impl.hpp

```diff
@@ -14,154 +15,94 @@
 
 // In case it hasn't been included yet.
 #include "save_image.hpp"
-#include "image_info.hpp"
 
 namespace mlpack {
 namespace data {
 
-/**
- * Save the given image to the given filename.
- *
- * @param filename Filename to save to.
- * @param matrix Matrix containing image to be saved.
- * @param info Information about the image (width/height/channels/etc.).
- * @param fatal Whether an exception should be thrown on save failure.
- */
+// Image saving API for multiple files.
 template<typename eT>
-bool Save(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal)
-{
-  arma::Mat<unsigned char> tmpMatrix =
-      arma::conv_to<arma::Mat<unsigned char>>::from(matrix);
+bool Save(const std::vector<std::string>& files,
+          const arma::Mat<eT>& matrix,
+          ImageOptions& opts)
 
-  return SaveImage(filename, tmpMatrix, info, fatal);
+{
+  return SaveImage(files, matrix, opts);
 }
 
-// Image saving API for multiple files.
 template<typename eT>
-bool Save(const std::vector<std::string>& files,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal)
+bool SaveImage(const std::vector<std::string>& files,
+               const arma::Mat<eT>& matrix,
+               ImageOptions& opts)
 {
-  if (files.size() == 0)
+  if (files.empty())
   {
-    if (fatal)
-    {
-      Log::Fatal << "Save(): vector of image files is empty; nothing to save."
-          << std::endl;
-    }
-    else
+    std::stringstream oss;
+    oss << "Save(): vector of image files is empty; nothing to save.";
+    HandleError(oss, opts);
+  }
+
+  // Check if we do have any type that is not supported.
+  for  (size_t i = 0; i < files.size() ; ++i)
+  {
+    if (!opts.saveType.count(Extension(files.at(i))))
```

Comment:
So what are the semantics on formats with `Save()`?  One nice thing about the explicit formats is that I can currently do

```
data::Save("my_file_with_no_extension", matrix, CSV);
```

and that will work just fine, since I specified the type as CSV.  Extension-based approaches are only used for `AutoDetect`.  So, for saving, I would imagine the same thing---if the user specifies `AutoDetect` or `Image`, then we detect the type of the image from the extension and check if it is supported for saving; otherwise, we use the explicit format that they gave and check that that is supported for saving (e.g. `PNG`, `JPG`, etc.).  Do you agree?  If so I guess this check needs to be modified depending on whether the user is auto-detecting or explicitly specified a format.

Reply:
Well, then, how do you detect if the user has specified this in advance?
We have to do a check based on extension in any case, since some files might have da ifferent extension, and not all of them might have the same extension